File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/VRAMVertex.java
package com.spvessel.spacevil;

import java.awt.Color;
import java.util.List;

import com.spvessel.spacevil.internal.Wrapper.OpenGLWrapper;
import static com.spvessel.spacevil.internal.Wrapper.OpenGLWrapper.*;

final class VramVertex extends AbstractVramResource {
    private float[] _vbo_data;
    int VBO;
    private int length;
    int type = GL_TRIANGLES;

    VramVertex() {
    }

    void sendColor(Shader shader, Color fill) {
        float[] argb = { (float) fill.getRed() / 255.0f, (float) fill.getGreen() / 255.0f,
                (float) fill.getBlue() / 255.0f, (float) fill.getAlpha() / 255.0f };
        sendUniform4f(shader, "background", argb);
    }

    void genBuffers(float[] vertices) {
        length = vertices.length / 2;
        // Vertices
        VBO = gl.GenBuffer();
        gl.BindBuffer(GL_ARRAY_BUFFER, VBO);
        gl.BufferDataf(GL_ARRAY_BUFFER, vertices.length * 4, vertices, GL_STATIC_DRAW);
        gl.VertexAttribPointer(0, 2, GL_FLOAT, false, 0, 0);
        gl.EnableVertexAttribArray(0);
    }

    void genBuffers(List<float[]> vertices) {
        length = vertices.size();
        // Vertices
        _vbo_data = new float[vertices.size() * 2];
        for (int i = 0; i < vertices.size(); i++) {
            int index = i * 2;
            _vbo_data[index + 0] = vertices.get(i)[0];
            _vbo_data[index + 1] = vertices.get(i)[1];
        }
        VBO = gl.GenBuffer();
        gl.BindBuffer(GL_ARRAY_BUFFER, VBO);
        gl.BufferDataf(GL_ARRAY_BUFFER, _vbo_data.length * 4, _vbo_data, GL_STATIC_DRAW);
        gl.VertexAttribPointer(0, 2, GL_FLOAT, false, 0, 0);
        gl.EnableVertexAttribArray(0);
    }
@Override
    public void draw() {
        gl.DrawArrays(type, 0, length);
        gl.DisableVertexAttribArray(0);
    }
@Override
    public void clear() {
        gl.DeleteBuffer(VBO);
        _vbo_data = null;
    }
@Override
    public void bind() {
        gl.BindBuffer(GL_ARRAY_BUFFER, VBO);
        gl.VertexAttribPointer(0, 2, GL_FLOAT, false, 0, 0);
        gl.EnableVertexAttribArray(0);
    }
@Override
    public void unbind() {
        gl.BindBuffer(GL_ARRAY_BUFFER, 0);
    }
}

File: SpaceVIL/Engine/VRAMVertex.cs
using System;
using System.Collections.Generic;
using System.Drawing;

using static OpenGL.OpenGLConstants;
using static OpenGL.OpenGLWrapper;

namespace SpaceVIL
{
    internal sealed class VramVertex : AbstractVramResource
    {
        private float[] _vbo_data;
        public uint[] VBO;
        private int length;
        internal uint Type = GL_TRIANGLES;

        internal VramVertex() { }

        internal void SendColor(Shader shader, Color fill)
        {
            float[] argb = { (float)fill.R / 255.0f, (float)fill.G / 255.0f, (float)fill.B / 255.0f, (float)fill.A / 255.0f };
            SendUniform4f(shader, "background", argb);
        }

        internal void GenBuffers(float[] vertices)
        {
            length = vertices.Length / 2;

            VBO = new uint[1];
            glGenBuffers(1, VBO);

            glBindBuffer(GL_ARRAY_BUFFER, VBO[0]);
            glBufferData(GL_ARRAY_BUFFER, vertices, GL_STATIC_DRAW);
            glVertexAttribPointer(0, 2, GL_FLOAT, false, 0, IntPtr.Zero);
            glEnableVertexAttribArray(0);
        }

        internal void GenBuffers(List<float[]> vertices)
        {
            length = vertices.Count;
            _vbo_data = new float[vertices.Count * 2];

            for (int i = 0; i < vertices.Count; i++)
            {
                int index = i * 2;
                _vbo_data[index + 0] = vertices[i][0];
                _vbo_data[index + 1] = vertices[i][1];
            }
            VBO = new uint[1];
            glGenBuffers(1, VBO);
            glBindBuffer(GL_ARRAY_BUFFER, VBO[0]);
            glBufferData(GL_ARRAY_BUFFER, _vbo_data, GL_STATIC_DRAW);
            glVertexAttribPointer(0, 2, GL_FLOAT, false, 0, IntPtr.Zero);
            glEnableVertexAttribArray(0);
        }

        public override void Draw()
        {
            glDrawArrays(Type, 0, length);
            glDisableVertexAttribArray(0);
        }

        public override void Clear()
        {
            glDeleteBuffers(1, VBO);
            _vbo_data = null;
        }

        public override void Bind()
        {
            glBindBuffer(GL_ARRAY_BUFFER, VBO[0]);
            glVertexAttribPointer(0, 2, GL_FLOAT, false, 0, IntPtr.Zero);
            glEnableVertexAttribArray(0);
        }

        public override void Unbind()
        {
            glBindBuffer(GL_ARRAY_BUFFER, 0);
        }
    }
}